Fix HTML AST JSON serializer for debugging

sosukesuzuki opened the pull request:
## Description

<!-- Please provide a brief summary of your changes: -->

Fixes #12609 

## Checklist

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

<!-- Please DO NOT remove the playground link -->

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

fisker commented:
Can you open another PR to try set `.parent` non-enumerable with this https://github.com/prettier/prettier/blob/f98c197b66db7491021f5d1cb65d137aa5d7ac68/src/language-html/ast.js#L180?

fisker commented:
Or maybe it should work if we add `.parent` to the prototype?

fisker commented:
If it doesn't work, we can also add `.toJSON` method. I think better fix in the html parser.

sosukesuzuki commented:
I've addressed your review with using `setNonEnumerableProperty` [38f59b0](https://github.com/prettier/prettier/pull/12611/commits/38f59b00e4f42b178805b5a007884ce2a0689830)

fisker commented on src/language-html/ast.js:
```diff
@@ -16,6 +16,8 @@ class Node {
       if (key in NODES_KEYS) {
         this._setNodes(key, value);
+      } else if (key === "parent") {
+        setNonEnumerableProperty(this, "parent", value);
```
I think this will break `Node#clone()`, but it seems already broken before since properties like `.prev` is also non-enumerable

sosukesuzuki replied:
Hmm, I just want to output the html AST with `--debug-print-ast` for other work, so what do you think about merging once the first implementation that specify the replacer function in the `JSON.stringify`?  I think other issues can be addressed later.

fisker commented:
There is another one in `cloneAndUpdateNodes`.
Let's fix this with https://github.com/prettier/prettier/pull/12611#issuecomment-1091162767? I think it's a better approach.

fisker commented:
Turns out setting `.parent` as non-enumerable did break the `clone()` and `new Node()`, but we can always clone `.parent` when calling `.clone()`.

- I added a new `.setProperty()` to replace the old `clone()`/`new()`/`_setNodes()` logic, I think it's more clear.
- I also turn `. attrMap` from non-enumerable property to a `getter`, so we don't need to copy it when cloning.

fisker commented:
Feel free to merge, if you are fine with those changes.

fisker commented on src/language-html/ast.js:
```diff
@@ -8,33 +8,40 @@ const NODES_KEYS = {
+    });
   }
 
   map(fn) {
```
This may need an update in future, seems overcomplicated.

sosukesuzuki commented:
Thanks
